File: frontend/src/api/aiAssistant.ts
/**
 * AI Story Creation Assistant API
 */

import client from './client';

// ========== Enums ==========

export type StoryGenre = 'murder_mystery' | 'thriller' | 'wuxia' | 'modern' | 'historical' | 'fantasy';
export type ClueImportance = 'high' | 'medium' | 'low';
export type GenerationStep = 'setting' | 'truth' | 'clue_chain' | 'npc_assignment' | 'detail_fill' | 'review';

// ========== Types ==========

export interface StorySettingInput {
  genre: StoryGenre;
  era: string;
  location: string;
  atmosphere?: string;
  npc_count: number;
  additional_notes?: string;
}

export interface TruthInput {
  murderer_hint?: string;
  motive_hint?: string;
  method_hint?: string;
  auto_generate?: boolean;
}

export interface TruthOption {
  murderer: string;
  motive: string;
  method: string;
  twist?: string;
  summary: string;
}

export interface TruthOptionsResponse {
  options: TruthOption[];
  recommendation_index: number;
  recommendation_reason?: string;
}

export interface SelectedTruth {
  murderer: string;
  motive: string;
  method: string;
  twist?: string;
}

export interface ClueNode {
  temp_id: string;
  name: string;
  importance: ClueImportance;
  description: string;
  reasoning_role: string;
  prereq_temp_ids: string[];
  suggested_npc_role?: string;
}

export interface ClueEdge {
  source: string;
  target: string;
}

export interface ClueChainValidation {
  is_valid: boolean;
  has_cycles: boolean;
  cycles: string[][];
  unreachable_clues: string[];
  root_clue_count: number;
  reasoning_path_count: number;
  warnings: string[];
}

export interface ClueChainSuggestion {
  nodes: ClueNode[];
  edges: ClueEdge[];
  reasoning_paths: string[][];
  validation: ClueChainValidation;
  ai_notes: string[];
}

export interface NPCKnowledgeScopePreview {
  knows: string[];
  does_not_know: string[];
  world_model_limits: string[];
}

export interface NPCSuggestion {
  temp_id: string;
  name: string;
  role: string;
  age?: number;
  background_summary: string;
  personality_traits: string[];
  assigned_clue_temp_ids: string[];
  knowledge_scope_preview: NPCKnowledgeScopePreview;
}

export interface NPCAssignmentResponse {
  npcs: NPCSuggestion[];
  unassigned_clue_temp_ids: string[];
  ai_notes: string[];
}

export interface ClueDetail {
  temp_id: string;
  name: string;
  detail: string;
  detail_for_npc: string;
  trigger_keywords: string[];
  trigger_semantic_summary: string;
}

export interface NPCDetail {
  temp_id: string;
  name: string;
  age?: number;
  background: string;
  personality: string;
  knowledge_scope: NPCKnowledgeScopePreview;
}

export interface DetailFillResponse {
  clue_details: ClueDetail[];
  npc_details: NPCDetail[];
  progress: number;
}

export interface StoryDraft {
  title: string;
  summary: string;
  background: string;
  difficulty: string;
  truth: Record<string, unknown>;
  clue_chain: ClueChainSuggestion;
  npcs: NPCSuggestion[];
  clue_details: ClueDetail[];
  npc_details: NPCDetail[];
  validation_result: ClueChainValidation;
  ready_to_create: boolean;
}

// ========== Request Types ==========

export interface GenerateTruthRequest {
  setting: StorySettingInput;
  hints?: TruthInput;
  llm_config_id?: string;
}

export interface GenerateClueChainRequest {
  setting: StorySettingInput;
  truth: SelectedTruth;
  existing_chain?: ClueChainSuggestion;
  llm_config_id?: string;
}

export interface GenerateNPCsRequest {
  setting: StorySettingInput;
  truth: SelectedTruth;
  clue_chain: ClueChainSuggestion;
  npc_count: number;
  llm_config_id?: string;
}

export interface GenerateDetailsRequest {
  setting: StorySettingInput;
  truth: SelectedTruth;
  clue_chain: ClueChainSuggestion;
  npcs: NPCSuggestion[];
  clue_temp_ids?: string[];
  llm_config_id?: string;
}

export interface OptimizeClueChainRequest {
  clue_chain: ClueChainSuggestion;
  focus?: string;
  llm_config_id?: string;
}

export interface CreateScriptResponse {
  script_id: string;
  title: string;
  message: string;
}

// ========== API Functions ==========

export const aiAssistantApi = {
  /**
   * Generate truth options based on story setting
   */
  generateTruth: async (request: GenerateTruthRequest): Promise<TruthOptionsResponse> => {
    const response = await client.post<TruthOptionsResponse>('/ai-assistant/generate-truth', request);
    return response.data;
  },

  /**
   * Generate clue chain from selected truth
   */
  generateClueChain: async (request: GenerateClueChainRequest): Promise<ClueChainSuggestion> => {
    const response = await client.post<ClueChainSuggestion>('/ai-assistant/generate-clue-chain', request);
    return response.data;
  },

  /**
   * Optimize existing clue chain
   */
  optimizeClueChain: async (request: OptimizeClueChainRequest): Promise<ClueChainSuggestion> => {
    const response = await client.post<ClueChainSuggestion>('/ai-assistant/optimize-clue-chain', request);
    return response.data;
  },

  /**
   * Generate NPCs and assign clues
   */
  generateNPCs: async (request: GenerateNPCsRequest): Promise<NPCAssignmentResponse> => {
    const response = await client.post<NPCAssignmentResponse>('/ai-assistant/generate-npcs', request);
    return response.data;
  },

  /**
   * Generate detailed content for clues and NPCs
   */
  generateDetails: async (request: GenerateDetailsRequest): Promise<DetailFillResponse> => {
    const response = await client.post<DetailFillResponse>('/ai-assistant/generate-details', request);
    return response.data;
  },

  /**
   * Create actual script from draft
   */
  createScript: async (draft: StoryDraft): Promise<CreateScriptResponse> => {
    const response = await client.post<CreateScriptResponse>('/ai-assistant/create-script', { draft });
    return response.data;
  },

  /**
   * Validate clue chain structure (no LLM)
   */
  validateClueChain: async (clueChain: ClueChainSuggestion): Promise<ClueChainValidation> => {
    const response = await client.post<ClueChainValidation>('/ai-assistant/validate-clue-chain', clueChain);
    return response.data;
  },
};

export default aiAssistantApi;

File: backend/app/api/ai_assistant.py
"""AI Story Creation Assistant API endpoints."""

import logging

from fastapi import APIRouter, HTTPException, status

from app.database import DBSession
from app.schemas.ai_assistant import (
    ClueChainSuggestion,
    CreateScriptFromDraftRequest,
    DetailFillResponse,
    GenerateClueChainRequest,
    GenerateDetailsRequest,
    GenerateNPCsRequest,
    GenerateTruthRequest,
    NPCAssignmentResponse,
    OptimizeClueChainRequest,
    StoryDraft,
    TruthOptionsResponse,
)
from app.services.ai_story_assistant import AIStoryAssistantService

logger = logging.getLogger(__name__)
router = APIRouter()


@router.post("/generate-truth", response_model=TruthOptionsResponse)
async def generate_truth_options(
    db: DBSession,
    request: GenerateTruthRequest,
) -> TruthOptionsResponse:
    """
    Generate multiple truth options based on story setting.

    AI will generate 3 different murder mystery plots for the user to choose from.

    Args:
        db: Database session.
        request: Story setting and optional hints.

    Returns:
        Multiple truth options with recommendations.
    """
    try:
        service = AIStoryAssistantService(db)
        return await service.generate_truth_options(request)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        logger.exception("Failed to generate truth options")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate truth options: {str(e)}",
        )


@router.post("/generate-clue-chain", response_model=ClueChainSuggestion)
async def generate_clue_chain(
    db: DBSession,
    request: GenerateClueChainRequest,
) -> ClueChainSuggestion:
    """
    Generate a clue chain from the selected truth.

    Uses reverse reasoning to build a logical clue dependency graph.

    Args:
        db: Database session.
        request: Story setting, selected truth, and optional existing chain.

    Returns:
        Clue chain with nodes, edges, and validation results.
    """
    try:
        service = AIStoryAssistantService(db)
        return await service.generate_clue_chain(request)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        logger.exception("Failed to generate clue chain")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate clue chain: {str(e)}",
        )


@router.post("/optimize-clue-chain", response_model=ClueChainSuggestion)
async def optimize_clue_chain(
    db: DBSession,
    request: OptimizeClueChainRequest,
) -> ClueChainSuggestion:
    """
    Optimize an existing clue chain.

    Fixes issues found in validation and improves the chain structure.

    Args:
        db: Database session.
        request: Existing clue chain and optimization focus.

    Returns:
        Optimized clue chain with updated validation.
    """
    try:
        service = AIStoryAssistantService(db)
        return await service.optimize_clue_chain(
            request.clue_chain,
            request.focus,
            request.llm_config_id,
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        logger.exception("Failed to optimize clue chain")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to optimize clue chain: {str(e)}",
        )


@router.post("/generate-npcs", response_model=NPCAssignmentResponse)
async def generate_npcs(
    db: DBSession,
    request: GenerateNPCsRequest,
) -> NPCAssignmentResponse:
    """
    Generate NPCs and assign clues to them.

    Creates NPCs that fit the story setting and assigns clues based on
    their backgrounds and the clue chain structure.

    Args:
        db: Database session.
        request: Story setting, truth, clue chain, and NPC count.

    Returns:
        List of NPCs with assigned clues.
    """
    try:
        service = AIStoryAssistantService(db)
        return await service.generate_npcs(request)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        logger.exception("Failed to generate NPCs")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate NPCs: {str(e)}",
        )


@router.post("/generate-details", response_model=DetailFillResponse)
async def generate_details(
    db: DBSession,
    request: GenerateDetailsRequest,
) -> DetailFillResponse:
    """
    Generate detailed content for clues and NPCs.

    Fills in the detail, detail_for_npc, trigger_keywords, etc. for clues,
    and full background, personality for NPCs.

    Args:
        db: Database session.
        request: All previous generation results.

    Returns:
        Detailed content for clues and NPCs.
    """
    try:
        service = AIStoryAssistantService(db)
        return await service.generate_details(request)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        logger.exception("Failed to generate details")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate details: {str(e)}",
        )


@router.post("/create-script", response_model=dict)
async def create_script_from_draft(
    db: DBSession,
    request: CreateScriptFromDraftRequest,
) -> dict:
    """
    Create actual Script, NPCs, and Clues from a story draft.

    This is the final step that persists the AI-generated content
    to the database as real entities.

    Args:
        db: Database session.
        request: Complete story draft.

    Returns:
        Created script ID and summary.
    """
    try:
        service = AIStoryAssistantService(db)
        script = await service.create_script_from_draft(request.draft)
        return {
            "script_id": script.id,
            "title": script.title,
            "message": "Script created successfully",
        }
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        )
    except Exception as e:
        logger.exception("Failed to create script from draft")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create script: {str(e)}",
        )


@router.post("/validate-clue-chain", response_model=dict)
async def validate_clue_chain(
    request: ClueChainSuggestion,
) -> dict:
    """
    Validate a clue chain without LLM.

    Performs structural validation on the clue chain graph.

    Args:
        request: Clue chain to validate.

    Returns:
        Validation results.
    """
    from app.services.ai_story_assistant import AIStoryAssistantService

    # Use static method for validation (no DB needed)
    validation = AIStoryAssistantService._validate_clue_chain(
        None,  # self not needed for this method
        request.nodes,
        request.edges,
    )

    return validation.model_dump()
